File: labs/class-11-clt/sampler.py
#!/usr/bin/env python3
"""
Understanding Uncertainty - Central Limit Theorem sampler (stdlib only)
-----------------------------------------------------------------------
The central limit theorem is a statement about the SAMPLE MEAN. Draw n
values with replacement from a population, average them, and you get one
sample mean X-bar. Repeat that m times and you have m sample means. This
script builds that "sampling distribution of the mean" and reports two
numbers the CLT predicts:

  mean_of_means -> should sit on the population mean  mu
  sd_of_means   -> should shrink like  sigma / sqrt(n)   (the standard error)

No numpy: every formula is spelled out. This mirrors sampler.js, which
carries out the same algorithm in JavaScript.

Usage:
  python3 sampler.py means [n] [m] [seed]   # print mean_of_means, sd_of_means
  python3 sampler.py draws [n] [m] [seed]   # print the m sample means, one/line
Defaults: n=30, m=2000, seed=1. The population is data.json next to this file.
"""
import sys, os, json, math, random


# ---- Summary statistics -------------------------------------
# mean and (population) standard deviation: sd divides by len, not len-1,
# so it matches pop_sd stored in data.json.
def mean(X):
    return sum(X) / len(X)


def std(X):
    m = mean(X)
    return math.sqrt(sum((x - m) ** 2 for x in X) / len(X))


# ---- The sampling distribution of the mean ------------------
# One "sample mean" is the average of n draws made WITH REPLACEMENT from
# the population X. rng.choice(X) draws one such value uniformly at random.
def one_sample_mean(X, n, rng):
    return sum(rng.choice(X) for _ in range(n)) / n


# Draw m independent sample means. Seeding random.Random(seed) makes the
# whole run reproducible: same seed -> same list of means.
def sample_means(X, n, m, seed):
    rng = random.Random(seed)
    return [one_sample_mean(X, n, rng) for _ in range(m)]


# ---- CLI ----------------------------------------------------
def main():
    here = os.path.dirname(os.path.abspath(__file__))
    with open(os.path.join(here, "data.json")) as f:
        data = json.load(f)
    X = data["values"]
    pop_sd = data["pop_sd"]

    argv = sys.argv[1:]
    cmd = argv[0] if argv else "means"
    n = int(argv[1]) if len(argv) > 1 else 30
    m = int(argv[2]) if len(argv) > 2 else 2000
    seed = int(argv[3]) if len(argv) > 3 else 1

    means = sample_means(X, n, m, seed)

    if cmd == "means":
        mom = mean(means)
        som = std(means)                     # spread of the sample means
        se_theory = pop_sd / math.sqrt(n)    # what the CLT predicts
        sys.stderr.write(f"# n={n}  m={m}  seed={seed}  pop_mean={mean(X):.4f}  pop_sd={pop_sd:.4f}\n")
        print(f"mean_of_means {mom:.6f}   (pop_mean = {mean(X):.6f})")
        print(f"sd_of_means   {som:.6f}   (sigma/sqrt(n) = {se_theory:.6f})")
    elif cmd == "draws":
        for v in means:
            print(f"{v:.6f}")
    else:
        sys.stderr.write("usage: python3 sampler.py [means|draws] [n] [m] [seed]\n")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: labs/class-11-clt/sampler.js
#!/usr/bin/env node
/* ============================================================
 * Understanding Uncertainty · Central Limit Theorem sampler
 * ------------------------------------------------------------
 * The central limit theorem is a statement about the SAMPLE MEAN.
 * Draw n values with replacement from a population, average them,
 * and you get one sample mean X-bar. Repeat m times and you have
 * the sampling distribution of the mean. This script reports the
 * two numbers the CLT predicts:
 *
 *   mean_of_means -> should sit on the population mean  mu
 *   sd_of_means   -> should shrink like  sigma / sqrt(n)  (std error)
 *
 * Pure stdlib JS, no deps, so you can read every line and compare
 * it to sampler.py (same commands, same math). The RNG here is the
 * shared makeLcg used by the lab widgets, so a given seed reproduces
 * exactly the numbers the page shows.
 *
 * Usage:
 *   node sampler.js means [n] [m] [seed]   # print mean_of_means, sd_of_means
 *   node sampler.js draws [n] [m] [seed]   # print the m sample means
 * Defaults: n=30, m=2000, seed=1. Population is data.json next door.
 * ============================================================ */

'use strict';

// ---- Reproducible RNG (same LCG as LabBase.makeLcg) ----------
function makeLcg(seed) {
  let s = seed >>> 0;
  return function () {
    s = (s * 1103515245 + 12345) & 0x7fffffff;
    return s / 0x7fffffff;
  };
}

// ---- Summary statistics -------------------------------------
// mean and (population) standard deviation: sd divides by length so it
// matches pop_sd stored in data.json.
function mean(X) { return X.reduce((a, b) => a + b, 0) / X.length; }
function std(X) {
  const m = mean(X);
  return Math.sqrt(X.reduce((a, b) => a + (b - m) * (b - m), 0) / X.length);
}

// ---- The sampling distribution of the mean ------------------
// One "sample mean" is the average of n draws made WITH REPLACEMENT.
function oneSampleMean(X, n, rand) {
  let s = 0;
  for (let i = 0; i < n; i++) s += X[Math.floor(rand() * X.length)];
  return s / n;
}

// Draw m independent sample means; seeding makes the run reproducible.
function sampleMeans(X, n, m, seed) {
  const rand = makeLcg(seed);
  const out = [];
  for (let j = 0; j < m; j++) out.push(oneSampleMean(X, n, rand));
  return out;
}

// ---- CLI -----------------------------------------------------
function main() {
  const fs = require('fs');
  const path = require('path');
  const data = JSON.parse(fs.readFileSync(path.join(__dirname, 'data.json'), 'utf8'));
  const X = data.values;
  const popSd = data.pop_sd;

  const cmd = process.argv[2] || 'means';
  const n = process.argv[3] ? parseInt(process.argv[3], 10) : 30;
  const m = process.argv[4] ? parseInt(process.argv[4], 10) : 2000;
  const seed = process.argv[5] ? parseInt(process.argv[5], 10) : 1;

  const means = sampleMeans(X, n, m, seed);

  if (cmd === 'means') {
    const mom = mean(means);
    const som = std(means);
    const seTheory = popSd / Math.sqrt(n);
    process.stderr.write(`# n=${n}  m=${m}  seed=${seed}  pop_mean=${mean(X).toFixed(4)}  pop_sd=${popSd.toFixed(4)}\n`);
    process.stdout.write(`mean_of_means ${mom.toFixed(6)}   (pop_mean = ${mean(X).toFixed(6)})\n`);
    process.stdout.write(`sd_of_means   ${som.toFixed(6)}   (sigma/sqrt(n) = ${seTheory.toFixed(6)})\n`);
  } else if (cmd === 'draws') {
    for (const v of means) process.stdout.write(`${v.toFixed(6)}\n`);
  } else {
    process.stderr.write('usage: node sampler.js [means|draws] [n] [m] [seed]\n');
    process.exit(1);
  }
}

if (typeof module !== 'undefined' && require.main === module) main();
if (typeof module !== 'undefined') {
  module.exports = { makeLcg, mean, std, oneSampleMean, sampleMeans };
}
